retry.go

```diff
@@ -41,6 +59,11 @@ func RetryNotify(operation Operation, b BackOff, notify Notify) error {
 			notify(err, next)
 		}
 
-		time.Sleep(next)
+		ch := time.After(next)
+		select {
+		case <-ctxDone:
```

Comment:
I think you should check for this (with an empty default case) before even calling `operation` for the first time.  It would surprise me a little to get the function called at all when passing in a canceled context.  I guess you could argue that the operation itself should also be checking the context if it's going to do anything with side effects.  I could go either way I guess.  Did you consider it?